server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexAliasesService.java

```diff
@@ -187,4 +189,13 @@ public ClusterState applyAliasActions(ClusterState currentState, Iterable<AliasA
         }
     }
 
+    private void validateAliasTargetIsNotDSBackingIndex(ClusterState currentState, AliasAction action) {
+        IndexAbstraction indexAbstraction = currentState.metadata().getIndicesLookup().get(action.getIndex());
+        assert indexAbstraction != null : "invalid cluster metadata. index [" + action.getIndex() + "] was not found";
+        if (indexAbstraction.getParentDataStream() != null) {
+            throw new IllegalArgumentException("The provided index [ " + action.getIndex()
+                + "] is a backing index belonging to data stream [" + indexAbstraction.getParentDataStream().getName()
+                + "]. data streams and their backing indices don't support aliases driven operations.");
```

Comment:
Minor error message suggestion -- I think "aliases driven operations" should be either "alias-driven operations" or just "alias operations." I prefer the latter as it's simple and clear.